src/openforms/submissions/form_logic.py

```diff
@@ -186,21 +188,22 @@ def evaluate_form_logic(
         # debounced client-side data changes
         data_diff = {}
 
-        for key, new_value in step.data.items():
-            original_value = initial_data.get(key)
+        for component in iter_components(configuration):
+            key = component["key"]
+            new_value = updated_step_data.get(key, empty)
+            original_value = initial_data.get(key, empty)
             # Reset the value of any field that may have become hidden again after evaluating the logic
-            if original_value:
+            if original_value is not empty:
                 component = get_component(configuration, key)
```

Comment:
this line is not required anymore, we already have the `component` from the for loop